src/server.c

```diff
@@ -2650,8 +2650,8 @@ void initServer(void) {
     /* Create the Redis databases, and initialize other internal state. */
     int slot_count_bits = (server.cluster_enabled) ? CLUSTER_SLOT_MASK_BITS : 0;
     for (j = 0; j < server.dbnum; j++) {
-        server.db[j].keys = kvstoreCreate(&dbDictType, slot_count_bits);
-        server.db[j].expires = kvstoreCreate(&dbExpiresDictType, slot_count_bits);
+        server.db[j].keys = kvstoreCreate(&dbDictType, slot_count_bits, KVSTORE_ALLOCATE_DICTS_ON_DEMAND);
+        server.db[j].expires = kvstoreCreate(&dbExpiresDictType, slot_count_bits, KVSTORE_ALLOCATE_DICTS_ON_DEMAND);
```

Comment:
@madolson please comment if you think we should drop this flag.
it means that a cluster node will only allocate a dict when the slot is assigned to it and populated, but on the other hand, when the slot is unassigned, the dict will remain allocated.
till now all 16k dicts were always allocated.
i'm worried that someone will find it odd.

p.s. maybe we can add a kvstore interface to explicitly release a specific empty dict, and call it when a slot is unassigned?

Reply:
@zuiderkwast maybe you wanna look into that?

Reply:
It's not that annoying. It's a bit weird that for static configurations there won't be the extra overhead, but there will be if you scale out and add move slots to other nodes. I don't think it's likely a big issue, and now that there is an interface we can improve it later.

Reply:
i'm bothered by two things:
1. after resharding there's wasted memory which we can reclaim.
2. the memory usage depends on the history and not just current state.

isn't it really simple to solve? some 10 trivial lines of code?

Reply:
i see madolson also reacted with a thumbs up emoji. i try to add it in #13072